File: mobile/refactor.dart
import 'dart:io';

void main() async {
  final file = File('lib/screens/category_screen.dart');
  String text = await file.readAsString();

  if (!text.contains('package:provider/provider.dart')) {
    text = text.replaceFirst(
      "import 'package:flutter/material.dart';",
      "import 'package:flutter/material.dart';\nimport 'package:provider/provider.dart';\nimport '../providers/category_provider.dart';"
    );
  }

  // Remove variables
  final varsToRemove = [
    RegExp(r'List<dynamic> products = \[\];\s*'),
    RegExp(r'bool isLoading = true;\s*'),
    RegExp(r'bool isFetchingMore = false;\s*'),
    RegExp(r'bool hasMore = true;\s*'),
    RegExp(r'String\? errorMessage;\s*'),
    RegExp(r'int currentPage = 1;\s*'),
    RegExp(r'List<dynamic> categories = \[\];\s*'),
    RegExp(r'int\? selectedCategoryId;\s*'),
    RegExp(r'String\? selectedCategoryName;\s*'),
    RegExp(r"String _sortBy = 'newest'; // newest \| price_asc \| price_desc \| popular\s*"),
    RegExp(r'RangeValues _priceRange = const RangeValues\(0, 50000000\);\s*'),
    RegExp(r'bool _filterInStock = false;\s*'),
    RegExp(r"String _searchQuery = '';\s*"),
  ];

  for (final r in varsToRemove) {
    text = text.replaceAll(r, '');
  }

  // Remove methods
  final methodsToRemove = [
    RegExp(r'Future<void> _loadAll\(\).*?^  }', multiLine: true, dotAll: true),
    RegExp(r'Future<void> fetchCategories\(\).*?^  }', multiLine: true, dotAll: true),
    RegExp(r'Future<void> fetchProducts\(\{.*?^  }', multiLine: true, dotAll: true),
    RegExp(r'void _resetAndFetch\(\).*?^  }', multiLine: true, dotAll: true),
    RegExp(r'void _onSearchChanged\(.*?^  }', multiLine: true, dotAll: true),
    RegExp(r'void _selectCategory\(.*?^  }', multiLine: true, dotAll: true),
  ];

  for (final r in methodsToRemove) {
    text = text.replaceAll(r, '');
  }

  // Build method
  text = text.replaceFirst(
    'Widget build(BuildContext context) {',
    'Widget build(BuildContext context) {\n    final provider = context.watch<CategoryProvider>();'
  );

  // Replacements
  final Map<String, String> wordReplacements = {
    'isLoading': 'provider.isLoading',
    'isFetchingMore': 'provider.isFetchingMore',
    'hasMore': 'provider.hasMore',
    'errorMessage': 'provider.errorMessage',
    'products': 'provider.products',
    'categories': 'provider.categories',
    'selectedCategoryId': 'provider.selectedCategoryId',
    'selectedCategoryName': 'provider.selectedCategoryName',
    '_sortBy': 'provider.sortBy',
    '_priceRange': 'provider.priceRange',
    '_filterInStock': 'provider.filterInStock',
    '_searchQuery': 'provider.searchQuery',
  };

  wordReplacements.forEach((key, value) {
    text = text.replaceAll(RegExp(r'\b' + key + r'\b'), value);
  });

  text = text.replaceAll('_resetAndFetch();', '');
  text = text.replaceAll('_selectCategory', 'provider.selectCategory');
  text = text.replaceAll('_onSearchChanged', 'provider.updateSearch');
  text = text.replaceAll('fetchProducts(loadMore: true)', 'provider.loadMore()');

  // Fix initState
  text = text.replaceAllMapped(
    RegExp(r'void initState\(\) \{.*?_loadAll\(\);.*?\}', multiLine: true, dotAll: true),
    (match) => '''void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      context.read<CategoryProvider>().loadAll();
    });
    _scrollCtrl.addListener(_onScroll);
  }'''
  );

  // Fix onScroll
  text = text.replaceAllMapped(
    RegExp(r'void _onScroll\(\) \{.*?\}', multiLine: true, dotAll: true),
    (match) => '''void _onScroll() {
    if (_scrollCtrl.position.pixels >= _scrollCtrl.position.maxScrollExtent - 250) {
      context.read<CategoryProvider>().loadMore();
    }
  }'''
  );

  // Fix apply filter
  text = text.replaceAllMapped(
    RegExp(r'setState\(\(\) \{\s*provider.sortBy = tmpSort;\s*provider.priceRange = tmpPrice;\s*provider.filterInStock = tmpInStock;\s*\}\);', multiLine: true, dotAll: true),
    (match) => 'context.read<CategoryProvider>().applyFilters(sort: tmpSort, price: tmpPrice, inStock: tmpInStock);'
  );

  // Fix hasActiveFilter
  text = text.replaceAll(
    'bool get _hasActiveFilter =>\n      provider.selectedCategoryId != null ||\n      provider.searchQuery.isNotEmpty ||\n      provider.filterInStock ||\n      provider.priceRange.start > 0 ||\n      provider.priceRange.end < 50000000 ||\n      provider.sortBy != \'newest\';',
    'bool get _hasActiveFilter => provider.selectedCategoryId != null || provider.searchQuery.isNotEmpty || provider.filterInStock || provider.priceRange.start > 0 || provider.priceRange.end < 50000000 || provider.sortBy != \'newest\';'
  );

  await File('lib/screens/category_screen_new.dart').writeAsString(text);
  print('Done');
}

File: mobile/refactor.py
import os, re

path = 'lib/screens/category_screen.dart'
with open(path, 'r', encoding='utf-8') as f:
    text = f.read()

if 'package:provider/provider.dart' not in text:
    text = text.replace('import \'package:flutter/material.dart\';', 'import \'package:flutter/material.dart\';\nimport \'package:provider/provider.dart\';\nimport \'../providers/category_provider.dart\';')

vars_to_remove = [
    r'List<dynamic> products = \[\];\s*',
    r'bool isLoading = true;\s*',
    r'bool isFetchingMore = false;\s*',
    r'bool hasMore = true;\s*',
    r'String\? errorMessage;\s*',
    r'int currentPage = 1;\s*',
    r'List<dynamic> categories = \[\];\s*',
    r'int\? selectedCategoryId;\s*',
    r'String\? selectedCategoryName;\s*',
    r'String _sortBy = \'newest\'; // newest \| price_asc \| price_desc \| popular\s*',
    r'RangeValues _priceRange = const RangeValues\(0, 50000000\);\s*',
    r'bool _filterInStock = false;\s*',
    r'String _searchQuery = \'\';\s*'
]
for v in vars_to_remove:
    text = re.sub(v, '', text)

methods_to_remove = [
    r'Future<void> _loadAll\(\).*?^  }',
    r'Future<void> fetchCategories\(\).*?^  }',
    r'Future<void> fetchProducts\(\{.*?^  }',
    r'void _resetAndFetch\(\).*?^  }',
    r'void _onSearchChanged\(.*?^  }',
    r'void _selectCategory\(.*?^  }',
]
for m in methods_to_remove:
    text = re.sub(m, '', text, flags=re.MULTILINE | re.DOTALL)

text = text.replace('Widget build(BuildContext context) {', 'Widget build(BuildContext context) {\n    final provider = context.watch<CategoryProvider>();')

replacements = {
    'isLoading': 'provider.isLoading',
    'isFetchingMore': 'provider.isFetchingMore',
    'hasMore': 'provider.hasMore',
    'errorMessage': 'provider.errorMessage',
    'products': 'provider.products',
    'categories': 'provider.categories',
    'selectedCategoryId': 'provider.selectedCategoryId',
    'selectedCategoryName': 'provider.selectedCategoryName',
    '_sortBy': 'provider.sortBy',
    '_priceRange': 'provider.priceRange',
    '_filterInStock': 'provider.filterInStock',
    '_searchQuery': 'provider.searchQuery',
}

def replace_word(word, replacement, txt):
    return re.sub(r'\b' + word + r'\b', replacement, txt)

for k, v in replacements.items():
    text = replace_word(k, v, text)

text = text.replace('fetchProducts(loadMore: true)', 'provider.loadMore()')
text = text.replace('_resetAndFetch();', '')
text = text.replace('_selectCategory', 'provider.selectCategory')
text = text.replace('_onSearchChanged', 'provider.updateSearch')

init_state_pattern = r'void initState\(\) \{.*?_loadAll\(\);.*?\}'
new_init = '''void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      context.read<CategoryProvider>().loadAll();
    });
    _scrollCtrl.addListener(_onScroll);
  }'''
text = re.sub(init_state_pattern, new_init, text, flags=re.MULTILINE | re.DOTALL)

on_scroll_pattern = r'void _onScroll\(\) \{.*?\}'
new_on_scroll = '''void _onScroll() {
    if (_scrollCtrl.position.pixels >= _scrollCtrl.position.maxScrollExtent - 250) {
      context.read<CategoryProvider>().loadMore();
    }
  }'''
text = re.sub(on_scroll_pattern, new_on_scroll, text, flags=re.MULTILINE | re.DOTALL)

apply_btn_pattern = r'setState\(\(\) \{\s*provider.sortBy = tmpSort;\s*provider.priceRange = tmpPrice;\s*provider.filterInStock = tmpInStock;\s*\}\);'
new_apply_btn = '''context.read<CategoryProvider>().applyFilters(sort: tmpSort, price: tmpPrice, inStock: tmpInStock);'''
text = re.sub(apply_btn_pattern, new_apply_btn, text, flags=re.MULTILINE | re.DOTALL)

with open('lib/screens/category_screen_new.dart', 'w', encoding='utf-8') as f:
    f.write(text)
print("Done")
